src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+
+using std::chrono::duration_cast;
+using Nanoseconds = std::chrono::nanoseconds;
+using Microseconds = std::chrono::microseconds;
+using Milliseconds = std::chrono::milliseconds;
+using Seconds = std::chrono::seconds;
+using Minutes = std::chrono::minutes;
+using Hours = std::chrono::hours;
+
+#if _HAS_CXX20
+    using Days = std::chrono::days;
+    using Weeks = std::chrono::weeks;
+    using Months = std::chrono::months;
+    using Years = std::chrono::years;
+#else
+    #include <ratio>
+    using Days = std::chrono::duration<int, std::ratio<86400>>;
+    using Weeks = std::chrono::duration<int, std::ratio<604800>>;
+    using Months = std::chrono::duration<int, std::ratio<2629746>>;
+    using Years = std::chrono::duration<int, std::ratio<31556952>>;
+#endif // _HAS_CXX20
```

Comment:
Yep, they should support fractional values. The underlying data type is std::chrono::duration --> https://en.cppreference.com/w/cpp/chrono/duration. 

These are just type aliases to durations with the calculated time conversions. I think the "baseline" is seconds so, for example, Weeks is the ration of seconds in a week. If you were to convert one nanosecond to weeks, it would be fractional because 1 nanosecond < 1 week. Does that make sense?